server_action_navigate/views/view_ir_actions_server.xml

```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright (C) 2020 - Today: GRAP (http://www.grap.coop)
+@author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+-->
+
+<odoo>
+
+    <record id="view_ir_actions_server_form" model="ir.ui.view">
+        <field name="model">ir.actions.server</field>
+        <field name="inherit_id" ref="base.view_server_action_form"/>
+
+        <field name="arch" type="xml">
+            <xpath expr="//page[@name='code']" position="after">
+                <page string="Navigation Options" name='page_navigate' autofocus="autofocus"
+                        attrs="{'invisible': [('state', '!=', 'navigate')]}">
+                    <group>
+                        <field name="max_navigate_line_sequence" invisible="1"/>
+                        <field name="max_navigate_line_model" invisible="1"/>
```

Comment:
all those hoops you jump through here made me think this must be simpler to implement, here's my proposal:

```diff
diff --git a/server_action_navigate/models/ir_actions_server_navigate_line.py b/server_action_navigate/models/ir_actions_server_navigate_line.py
index fab5326..c56731e 100644
--- a/server_action_navigate/models/ir_actions_server_navigate_line.py
+++ b/server_action_navigate/models/ir_actions_server_navigate_line.py
@@ -2,7 +2,7 @@
 # @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from odoo import fields, models
+from odoo import api, fields, models
 
 
 class IrActionsServerNavigateLine(models.Model):
@@ -10,14 +10,7 @@ class IrActionsServerNavigateLine(models.Model):
     _description = "Server Actions Navigation Lines"
     _order = "sequence"
 
-    sequence = fields.Integer(string="Sequence", default=9999)
-
-    sequence2 = fields.Integer(
-        related='sequence',
-        string="Line Number",
-        readonly=True,
-        store=True,
-    )
+    sequence = fields.Integer(string="Sequence", default=1)
 
     field_model = fields.Char(
         string="Model", related="field_id.relation", store=True)
@@ -29,3 +22,23 @@ class IrActionsServerNavigateLine(models.Model):
     field_id = fields.Many2one(
         comodel_name="ir.model.fields", string="Field",
         required=True)
+
+    # when adding a record, onchange is called for every field on the
+    # form, also in editable list views
+    @api.onchange('field_id')
+    def _onchange_field_id(self):
+        # check out the docstring of this in odoo/models.py
+        lines = self.action_id.resolve_2many_commands(
+            'navigate_line_ids',
+            self.env.context.get('navigate_line_ids', []),
+        )
+        lines = sum(map(self.new, lines), self.browse([]))
+        model = lines[-1:].field_id.relation or self.action_id.model_id.model
+        return {
+            'domain': {
+                'field_id': [
+                    ('ttype', 'in', ['many2one', 'one2many', 'many2many']),
+                    ('model', '=', model),
+                ],
+            }
+        }
diff --git a/server_action_navigate/views/view_ir_actions_server.xml b/server_action_navigate/views/view_ir_actions_server.xml
index a08b9af..7611479 100644
--- a/server_action_navigate/views/view_ir_actions_server.xml
+++ b/server_action_navigate/views/view_ir_actions_server.xml
@@ -18,21 +18,19 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
                     <group>
                         <field name="max_navigate_line_sequence" invisible="1"/>
                         <field name="max_navigate_line_model" invisible="1"/>
+                        <!-- we can check emptyness of a 2many like this since v10 i think /-->
                         <button name="delete_last_line" string="Delete Last Line" type="object"
-                            attrs="{'invisible': [('max_navigate_line_sequence', '=', 1)]}"/>
-                        <field name="navigate_line_ids" context="{'default_sequence': max_navigate_line_sequence}" nolabel="1" colspan="4">
+                            attrs="{'invisible': [('navigate_line_ids', '=', [])]}"/>
+                        <field name="navigate_line_ids" context="{'navigate_line_ids': navigate_line_ids}" nolabel="1" colspan="4">
                             <tree editable="bottom" delete="false">
-                                <field name="sequence" invisible="1"/>
-                                <field name="sequence2"/>
-                                <field name="field_id" domain="[
-                                    ('model', '=', parent.max_navigate_line_model),
-                                    ('ttype', 'in', ['many2one', 'one2many', 'many2many']),
-                                    ]"/>
+                                <!-- the handle widget takes care of the sequence /-->
+                                <field name="sequence" widget="handle" invisible="True" />
+                                <field name="field_id" />
                                 <field name="field_model"/>
                             </tree>
                         </field>
                         <field name="navigate_action_id"
-                            attrs="{'invisible': [('max_navigate_line_sequence', '=', 1)]}"/>
+                            attrs="{'invisible': [('navigate_line_ids', '=', [])]}"/>
                     </group>
                 </page>
             </xpath>
```

I don't have my ssh keys right now, otherwise I'd just PR this. You can then delete all the UI helper fields and the code for them above.

Personally, I'd also allow deletions and just add constraint that forbids to make the order inconsistent. This way you stick to the Odoo UI way of doing stuff.